### PR TITLE
Rake task to build term generation report

### PR DESCRIPTION
**Pull request by whoojemaflip**
Rake task and calculator function to generate and save a CSV for the "rate of return of page review"

[Trello](https://trello.com/c/x6GUpctA/173-generate-graph-showing-rate-of-term-generation-for-coming-to-the-uk)

### CODE REVIEW

File: lib/tasks/term_generation/calculate_project_statistics.rake [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    terms = Set.new([])
+    project.taxonomy_todos.done.order(completed_at: :asc).each do |todo|
+      count_of_previously_found_terms = terms.size
+      terms.merge Set.new(todo.terms)
+      todo.update_attribute(:count_of_new_terms, terms.size - count_of_previously_found_terms)
+
+      previous50 = TaxonomyTodo.done
+                                .order(completed_at: :desc)
+                                .where("completed_at <= ?", todo.completed_at)
+                                .limit(50)
```
Comment (koetsier) [inline lib/tasks/term_generation/calculate_project_statistics.rake]: could you make the '50' a constant/parameter? It seems a bit hard-coded like this.

File: lib/tasks/term_generation/calculate_project_statistics.rake [hunk 2]
```diff
@@ -0,0 +1,35 @@
+namespace :term_generation do
+  desc "Calculate statistics for a Term Generation project"
+  task :calculate_project_statistics, [:project_id] => :environment do |_, args|
+    project = TaxonomyProject.find(args.project_id)
```
Comment (koetsier) [inline lib/tasks/term_generation/calculate_project_statistics.rake]: You could extract the logic of updating the statistics to a separate service object or maybe the model itself so you can write unit tests

File: lib/tasks/term_generation/calculate_project_statistics.rake (hunk 1 above)
Comment (whoojemaflip) [inline lib/tasks/term_generation/calculate_project_statistics.rake]: I don't think it makes any sense as a parameter, unless we intend to build a generic "moving window average" algorithm.

File: app/domain/term_generation/rate_of_discovery.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Calculate the moving-window average of the absolute change
+    # in new term discovery for a project. Aims to give insight
+    # into the expected reward of reviewing a single extra page.
+
+    def self.calculate(project)
+      results = []
+
+      # Keeps track of previously seen unique terms.
+      terms = Set.new([])
```
Comment (koetsier) [inline app/domain/term_generation/rate_of_discovery.rb]: can you not just call Set.new()?

Comment (whoojemaflip) [inline app/domain/term_generation/rate_of_discovery.rb]: updated

File: app/domain/term_generation/rate_of_discovery.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        # Add the terms from this Todo to the set of all terms.
+        terms.merge Set.new(todo.terms)
+
+        # Record the absolute change in number of terms encountered for this Todo.
+        results << {
+          count_of_new_terms: terms.size - count_of_previously_found_terms
+        }
+
+        # Our window is the last 50 results.
+        moving_window = results.reverse[0..WINDOW_SIZE - 1]
```
Comment (koetsier) [inline app/domain/term_generation/rate_of_discovery.rb]: can you not use results.last(WINDOW_SIZE)?

Comment (whoojemaflip) [inline app/domain/term_generation/rate_of_discovery.rb]: I had a total blind spot for `Enum#last` thanks for pointing that out!

File: app/domain/term_generation/rate_of_discovery.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+        # Calculate the absolute change in terms found over the course of the window.
+        total_change = moving_window.sum { |r| r[:count_of_new_terms] }
+
+        # Average the result. Expected values are roughly between 0.3 and 1.5
+        results[-1][:moving_average] = (total_change * 1.0) / moving_window.size
```
Comment (koetsier) [inline app/domain/term_generation/rate_of_discovery.rb]: is there an advantage over total_change.to_f?

Comment (whoojemaflip) [inline app/domain/term_generation/rate_of_discovery.rb]: Force of habit. Updated.

**Pull request merged**